independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java

```diff
@@ -209,8 +228,42 @@ void notify(T event) {
         }
 
         void notify(T event, ObserverExceptionHandler exceptionHandler, boolean async) {
-            if (!isEmpty()) {
+            if (!transactionalObserverMethodsEmpty()) {
+                InstanceHandle<TransactionSynchronizationRegistry> registryInstance = Arc.container()
+                        .instance(TransactionSynchronizationRegistry.class);
+                if (registryInstance.isAvailable() && registryInstance.get().getTransactionStatus() == 0) {
+                    // we have one or more transactional OM, and TransactionSynchronizationRegistry is available
+                    // we attempts to register JTA synchronization
+                    List<DeferredEventNotification<?>> deferredEvents = new ArrayList<>();
+                    EventContext eventContext = new EventContextImpl<>(event, eventMetadata);
+                    for (ObserverMethod<? super T> om : transactionalObserverMethods) {
+                        deferredEvents
+                                .add(new DeferredEventNotification(om, eventContext, Status.valueOf(om.getTransactionPhase())));
+                    }
+                    Synchronization sync = new ArcSynchronization(deferredEvents);
+                    TransactionSynchronizationRegistry registry = registryInstance.get();
+                    try {
+                        registry.registerInterposedSynchronization(sync);
+                    } catch (Exception e) {
+                        if (e.getCause() instanceof RollbackException || e.getCause() instanceof IllegalStateException) {
+                            // registration failed, merge back with non-transactional methods
+                            // AFTER_SUCCESS are accordingly to CDI spec left out
+                            observerMethods.addAll(transactionalObserverMethods.stream()
+                                    .filter(observerMethod -> !observerMethod.getTransactionPhase()
+                                            .equals(TransactionPhase.AFTER_SUCCESS))
+                                    .collect(Collectors.toList()));
+                        }
+                    }
+                } else {
+                    // no JTA available, we will have to notify transactional OM along with all other OM
+                    observerMethods.addAll(transactionalObserverMethods);
+                    // TODO log this somehow?
```

Comment:
Not sure if we should log this somehow - we know there are tx observers but either Narayna isn't present (no access to SyncRegistry) or no tx is in progress ATM.